torch/_inductor/scheduler.py

```diff
@@ -3951,11 +3960,174 @@ def can_buffer_be_removed_through_fusion(
+                unmet_output_names - returned_output_names
+            )
+
+        return signatures[::-1]
+
+    def graph_partition(
+        self,
+    ) -> tuple[list[PartitionType], list[GraphPartitionSignature]]:
+        """
+        Given a list of BaseSchedulerNodes, split into a list of
+        graph partitions and compute partition input/output signatures.
+        """
+        partitions: list[PartitionType] = []
+
+        cur_partition: PartitionType = []
+        for node in self.nodes:
+            if self.should_partition(node):
+                if cur_partition:
+                    partitions.append(cur_partition)
+                    cur_partition = []
+                partitions.append([node])
+            else:
+                cur_partition.append(node)
+
+        if cur_partition:
+            partitions.append(cur_partition)
+
+        return partitions, self.get_graph_partition_signature(partitions)
+
     def codegen(self) -> None:
         with dynamo_timed("Scheduler.codegen"):
-            return self._codegen()
+            return (
+                self._codegen_partitions()
+                if torch._inductor.config.graph_partition
+                and not V.graph.cpp_wrapper
+                and not V.graph.aot_mode
+                else self._codegen(self.nodes)
+            )
+
+    def _codegen_partition_wrapper(
+        self,
+        partition: PartitionType,
+        signature: GraphPartitionSignature,
+    ) -> None:
+        """Codegen a partition given its inputs/outputs"""
+        parent_wrapper_code = V.graph.wrapper_code
+        graph_partition_id = next(self._graph_partition_counter)
+
+        with V.graph.set_current_wrapper_code():
+            V.graph.init_wrapper_code(
+                is_subgraph=True,
+                subgraph_name=f"partition_{graph_partition_id}",
+                parent_wrapper_code=parent_wrapper_code,
+                partition_signatures=signature,
+            )
+            self._codegen(partition)
+            partition_code, _ = V.graph.wrapper_code.generate(V.graph.is_inference)
+
+        V.graph.wrapper_code.define_subgraph_launcher_fn(partition_code)
+
+        V.graph.wrapper_code.codegen_partition_call(graph_partition_id, signature)
+        V.graph.wrapper_code.allocated.update(
+            [node.get_name() for node in signature.output_nodes]
+        )
 
-    def _codegen(self) -> None:
+    def _codegen_partitions(self) -> None:
+        partitions, signatures = self.graph_partition()
+
+        for partition, signature in zip(partitions, signatures):
+            assert (
+                len(partition) >= 1
+            ), f"Each partition must have at least one node but found {len(partition)}"
+
+            if len(partition) > 1:
+                self._codegen_partition_wrapper(partition, signature)
+            else:
+                # Inline small partitions to avoid overheads from function calls.
+                self._codegen(partition)
```

Comment:
to start, i might consider just codegening everyhting. because this gets into a policy decision where we are not codegening single node cudagraphs. 

that might be what we want but i would separate out the policy decision elsewhere.. traditionally, we cudagraph as much as we can because of memory concerns.

this is another case where we might want to know if the partition is meant to be cudagraphd or not. although i think in your subsequent metadata pr that will be more evident.